File: src/kvm-book/scripts/libvirt-dominfo.c
/**
 * Get domain information via libvirt C API.
 * Tested with python 2.6 and libvirt-devel-0.9.10 on a RHEL 6.3 host system.
 */

#include <stdio.h>
#include <libvirt/libvirt.h>

int getDomainInfo(int id) {
	virConnectPtr conn = NULL; /* the hypervisor connection */
	virDomainPtr dom = NULL;   /* the domain being checked */
	virDomainInfo info;        /* the information being fetched */

	/* NULL means connect to local QEMU/KVM hypervisor */
	conn = virConnectOpenReadOnly(NULL);
	if (conn == NULL) {
		fprintf(stderr, "Failed to connect to hypervisor\n");
		return 1;
	}

	/* Find the domain by its ID */
	dom = virDomainLookupByID(conn, id);
	if (dom == NULL) {
		fprintf(stderr, "Failed to find Domain %d\n", id);
		virConnectClose(conn);
		return 1;
	}

	/* Get virDomainInfo structure of the domain */
	if (virDomainGetInfo(dom, &info) < 0) {
		fprintf(stderr, "Failed to get information for Domain %d\n", id);
		virDomainFree(dom);
		virConnectClose(conn);
		return 1;
	}

	/* Print some info of the domain */
	printf("Domain ID: %d\n", id);
	printf("    vCPUs: %d\n", info.nrVirtCpu);
	printf("   maxMem: %d KB\n", info.maxMem);
	printf("   memory: %d KB\n", info.memory);

	if (dom != NULL)
		virDomainFree(dom);
	if (conn != NULL)
		virConnectClose(conn);
	
	return 0;
}

int main(int argc, char **argv)
{
	int dom_id = 3;
	printf("-----Get domain info by ID via libvirt C API -----\n");
	getDomainInfo(dom_id);
	return 0;
}

File: src/kvm-book/scripts/libvirt-dominfo.py
#!/usr/bin/python
# Get domain info via libvirt python API.
# Tested with python2.6 and libvirt-python-0.9.10 on KVM host.

import libvirt
import sys

def createConnection():
    conn = libvirt.openReadOnly(None)
    if conn == None:
        print 'Failed to open connection to QEMU/KVM'
        sys.exit(1)
    else:
	print '-----Connection is created successfully-----'
        return conn

def closeConnnection(conn):
    print ''
    try:
        conn.close()
    except:
        print 'Failed to close the connection'
        return 1

    print 'Connection is closed'

def getDomInfoByName(conn, name):
    print ''
    print '----------- get domain info by name ----------"'
    try:
        myDom = conn.lookupByName(name)
    except:
        print 'Failed to find the domain with name "%s"' % name
        return 1

    print "Dom id: %d   name: %s" % (myDom.ID(), myDom.name())
    print "Dom state: %s" % myDom.state(0)
    print "Dom info: %s" % myDom.info()
    print "memory: %d MB" % (myDom.maxMemory()/1024)
    print "memory status: %s" % myDom.memoryStats()
    print "vCPUs: %d" % myDom.maxVcpus()

def getDomInfoByID(conn, id):
    print ''
    print '----------- get domain info by ID ----------"'
    try:
        myDom = conn.lookupByID(id)
    except:
        print 'Failed to find the domain with ID "%d"' % id
        return 1

    print "Domain id is %d ; Name is %s" % (myDom.ID(), myDom.name())

if __name__ == '__main__':
    name1 = "kvm-guest"
    name2 = "notExist"
    id1 = 3
    id2 = 9999
    print "---Get domain info via libvirt python API---"
    conn = createConnection()
    getDomInfoByName(conn, name1)
    getDomInfoByName(conn, name2)
    getDomInfoByID(conn, id1)
    getDomInfoByID(conn, id2)
    closeConnnection(conn)
